README.md
# msp430-cli

A Command Line Interface (CLI) for MSP430G2 LaunchPad

## Environment

* PC:        Kubuntu-18.04
* Compiler:  msp430-elf-gcc
* Flasher:   mspdebug
* IDE:       None
* Build:     Makefile
* Serial Console: screen

## How to compile and upload the program

To compile and load program into the board, you need to install some requisite tools such as: *msp430-elf-gcc*, *mspdebug* and *make*.
After above tools have already installed completely, you perform the guide as follows:

* Compile and upload the program on MCU

  ```shell
  make all
  ```

* Debug (if needed)

  * Start gdb server with port 2000

    ```shell
    sudo mspdebug rf2500
    ```

  * Another windows, start gdb client

    ```shell
    make debug
    ```

  * Finally, connect to gdb server and debug.

    ```shell
    (gdb) target remote localhost:2000
    (gdb) load build/cli_main.elf
    (gdb) continue
    ```

* Clean the project

  ```shell
  make clean
  ```

* Using the *screen* for serial interface

  ```shell
  sudo screen /dev/ttyACM0 -b 9600
  ```

## Screenshot

![MSP430-CLI](https://raw.githubusercontent.com/nhivp/msp430-cli/master/docs/imgs/msp430-cli.png "MSP430-CLI")

## Changelog

See [CHANGELOG.md](CHANGELOG.md)

## Useful links

* [A list of embedded command-line-interface projects](http://www.dalescott.net/an-embedded-command-line-interface/)

## Contributing
Feel free to create pull requests and I will gladly accept them! :clap: :clap:

Good luck, and happy hacking?! :yum:


CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.

## [0.0.1](https://github.com/nhivp/msp430-cli/releases/tag/v0.1) - 2018/08/18

### Bug Fixes

* None

### Code Refactoring

* None

### Features

* Check the used ram/flash
* Perform some basic command: help, info, ...
* First implement


inc/msp_uart.h
/**
 * Copyright (C) 2018  nhivp
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#ifndef _MSP_UART_H_
#define _MSP_UART_H_
#include "msp430g2553.h"

#define UART_RX_VECTOR  USCIAB0RX_VECTOR

extern void uart_init(void);
extern void uart_putc(unsigned char character);
extern void uart_puts(char* s);
extern unsigned char uart_getc(void);

#endif /* _MSP_UART_H_ */


src/msp_uart.c
/**
 * Copyright (C) 2018  nhivp
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include "msp_uart.h"

/**
 * Initialize UART
 */
void uart_init(void)
{
    /* Configure hardware UART */
    P1SEL = BIT1 + BIT2 ;  // P1.1 = RXD, P1.2=TXD
    P1SEL2 = BIT1 + BIT2 ; // P1.1 = RXD, P1.2=TXD
    UCA0CTL1 |= UCSSEL_2;  // Use SMCLK
    UCA0BR0 = 104;         // Set baud rate to 9600 with 1MHz clock (Data Sheet 15.3.13)
    UCA0BR1 = 0;           // Set baud rate to 9600 with 1MHz clock
    UCA0MCTL = UCBRS0;     // Modulation UCBRSx = 1
    UCA0CTL1 &= ~UCSWRST;  // Initialize USCI state machine
    IE2 |= UCA0RXIE;       // Enable USCI_A0 RX interrupt
}

/**
 * Send a single byte out through UART
 */
void uart_putc(unsigned char character)
{
    while (!(IFG2 & UCA0TXIFG)); // USCI_A0 TX buffer ready?
    UCA0TXBUF = character;       // TX -> RXed character
}

/**
 * Sends a string out through UART
 */
void uart_puts(char* s)
{
    while (*s != '\0')
    {
        uart_putc(*s);
        s++;
    }
}

/**
 * Receive a single byte out through UART
 */
unsigned char uart_getc(void)
{
    return UCA0RXBUF;
}


src/msp_printf.c
/**
 * Copyright (C) 2018  nhivp
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include "msp_printf.h"
#include "stdarg.h"

static void xtoa(unsigned long x, int width, const int *wp, const unsigned long *dp);
static void puth(unsigned n);

static const unsigned long dv[] =
{
    // 4294967296 // 32 bit unsigned max
    1000000000, // +0
    100000000, // +1
    10000000, // +2
    1000000, // +3
    100000, // +4
    // 65535 // 16 bit unsigned max
    10000, // +5
    1000, // +6
    100, // +7
    10, // +8
    1, // +9
};

static const int dw[] =
{ // Width of integer
    // 4294967296 // 32 bit unsigned max
    10, // +0
    9, // +1
    8, // +2
    7, // +3
    6, // +4
    // 65535 // 16 bit unsigned max
    5, // +5
    4, // +6
    3, // +7
    2, // +8
    1, // +9
};

static void xtoa(unsigned long x, int width, const int *wp, const unsigned long *dp)
{
    char c;
    unsigned long d;

    if(x)
    {
        while(x < *dp)
        {
            ++dp;
            if (width >= *wp) uart_putc('0');
            ++wp;
        }

        do
        {
            d = *dp++;
            c = '0';
            while(x >= d) ++c, x -= d;
            uart_putc(c);
        } while(!(d & 1));
    }
    else
    {
        while (width--) uart_putc('0');
    }
}

static void puth(unsigned n)
{
    static const char hex[16] = { '0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'};
    uart_putc(hex[n & 15]);
}

void printformat(char *format, ...)
{
    char formatChar;
    int i;
    long n;
    int width = 1;

    va_list a;
    va_start(a, format);
    while(*format)
    {
        formatChar = *format++;
        if(formatChar == '%')
        {
            formatChar = *format++;
            if ((formatChar > 48) && (formatChar < 58) )
            {
                width = formatChar-48;
                formatChar = *format++;
            }
            switch(formatChar)
            {
                case 's': // String
                    uart_puts(va_arg(a, char*));
                    break;
                case 'c': // Char
                    uart_putc(va_arg(a, unsigned));
                    break;
                case 'i': // 16 bit Integer
                case 'u': // 16 bit Unsigned
                    i = va_arg(a, int);
                    if(formatChar == 'i' && i < 0) i = -i, uart_putc('-');
                    xtoa((unsigned)i, width, dw + 5, dv + 5); // Send 16 bit integer to xtoa and point to +5 element of dv
                    break;
                case 'l': // 32 bit Long
                case 'n': // 32 bit unsigned long
                    n = va_arg(a, long);
                    if(formatChar == 'l' && n < 0) n = -n, uart_putc('-');
                    xtoa((unsigned long)n, width, dw, dv);
                    break;
                case 'x': // 16 bit hexadecimal
                    i = va_arg(a, int);
                    puth(i >> 12);
                    puth(i >> 8);
                    puth(i >> 4);
                    puth(i);
                    break;
                case 0: return;
                default: goto bad_fmt;
            }
        }
        else
            bad_fmt: uart_putc(formatChar);
    }
    va_end(a);
}


src/main.c
/**
 * Copyright (C) 2018  nhivp
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include "msp_printf.h"
#include "msp_uart.h"
#include <stdint.h>
#include <stdbool.h>
#include <string.h>

/**
 *  Typedef definition
 */
typedef void (*CLI_Command_Function_t)(void);

typedef struct
{
    char Command[32];
    char Command_Desc[64];
    CLI_Command_Function_t Command_Func;
} CLI_Command_t;

/**
 *  Macro definition
 */
#define COMMAND_LEN(x)            sizeof(x)/sizeof(*(&x[0]))
#define COMMAND_STRING_LEN        128

/**
 * Private function prototypes
 */
static void board_init(void);
static void startup_cli(void);
static void CLI_GetCommand(unsigned char* cmd);
static void CLI_Help(void);
static void CLI_Hello(void);
static void CLI_Info(void);

/**
 *  Private variables
 */
static CLI_Command_t command_tbl[] =
{
    /* Command, Description,                    Command_Func */
    { "help"  , "Show a list of command.",      CLI_Help     },
    { "info"  , "Show all features of MCU",     CLI_Info     },
    { "hello" , "Say Hello, World.",            CLI_Hello    }
};

/**
 *  Exported variables
 */
unsigned char parameterString[COMMAND_STRING_LEN];
uint8_t parameterLength;
volatile bool validCommandFlag;
extern volatile uint16_t __m_flash_size;
extern volatile uint16_t __m_ram_size;

/**
 * Main App
 */
int main(void)
{
    unsigned char cmd[32];
    uint8_t cmd_idx;

    board_init();
    uart_init();
    memset(parameterString, '\0', COMMAND_STRING_LEN);
    memset(cmd, '\0', 32);

    /* Show banner */
    startup_cli();
    /* Enable interrupt */
    __bis_SR_register(GIE);

    while (1)
    {
        /* The 'validCommandFlag' is true when the user enter an input command from console */
        while (validCommandFlag)
        {
            CLI_GetCommand(cmd);

            if (cmd[0] == '\0')
            {
                printf("\r\nMissing command!");
            }
            else
            {
                for (cmd_idx = 0; cmd_idx < COMMAND_LEN(command_tbl); cmd_idx++)
                {
                    if (!strcmp((char*)cmd, (char*)command_tbl[cmd_idx].Command))
                    {
                        break;
                    }
                }

                if (cmd_idx < COMMAND_LEN(command_tbl))
                {
                    /* Execute command */
                    command_tbl[cmd_idx].Command_Func();
                }
                else
                {
                    printf("\r\nInvalid command!");
                }
            }

            printf("\r\nMSP430@CLI > ");

            /* Reset receive buffer and flag*/
            memset(parameterString, '\0', parameterLength + 1);
            memset(cmd, '\0', 32);
            parameterLength = 0;
            validCommandFlag = false;
        }
    }
}

/**
 * Initialize the board MSP430G2 LaunchPad
 */
static void board_init(void)
{
    WDTCTL = WDTPW + WDTHOLD; // Stop WDT
    /* Load calibrated DCO vlues to set CPU clock to 1MHz */
    BCSCTL1 = CALBC1_1MHZ;    // Set DCO to 1MHz
    DCOCTL = CALDCO_1MHZ;     // Set DCO to 1MHz
}

/**
 * Start CLI
 */
static void startup_cli(void)
{
    printf("\r\n");
    printf("*----------------------------------------*\r\n");
    printf("*         MSP-EXP430G2 LaunchPad         *\r\n");
    printf("*         Command Line Interface         *\r\n");
    printf("*----------------------------------------*\r\n");
    printf("\n<< System Info >>\r\n");
    printf("\tMCU:                     MSP430G2553\r\n");
    printf("\tFLASH:                   16kB\r\n");
    printf("\tFLASH used:              %u\r\n", __m_flash_size);
    printf("\tRAM:                     512B\r\n");
    printf("\tRAM used:                %u\r\n", __m_ram_size);
    printf("\tSystem clock:            1MHz\r\n");
    printf("\tSystem console baudrate: 9600bps\r\n");
    printf("\r\n\r\nMSP430@CLI > ");
}

/**
 *  Get a command from the input string.
 */
static void CLI_GetCommand(unsigned char* cmd)
{
    uint8_t cmd_len;

    for (cmd_len = 0; cmd_len < parameterLength; cmd_len++)
    {
        if ((parameterString[cmd_len] == ' ')
            || (parameterString[cmd_len] == '\0')
            || (parameterString[cmd_len] == 0x0D))
        {
            strncpy((char*)cmd, (char*)parameterString, cmd_len);
            break;
        }
    }
}

/**
 * Command executing: help
 */
static void CLI_Help(void)
{
    uint8_t i;
    /* Print all commands and description for usage */
    printf( "\r\nPlease input command as follows:");

    for (i = 0; i < COMMAND_LEN(command_tbl); i++)
    {
        printf("\r\n\t%s: %s", command_tbl[i].Command, command_tbl[i].Command_Desc);
    }
    printf("\r\n");
}

/**
 * Command executing: hello
 */
static void CLI_Hello(void)
{
    /* Say "Hello, World!"" */
    printf("\r\nHello, World!");
    printf("\r\nI'm Nhi Pham. You'll find me on Earth.");
}

/**
 * Command executing: info
 */
static void CLI_Info(void)
{
    printf("\r\n<< Device Info >>");
    printf("\r\n\tCPU:             msp430g2553 @ 16bit RISC Architecture");
    printf("\r\n\tArchitecture:    16bit RISC Architecture");
    printf("\r\n\tCPU clock:       16MHz");
    printf("\r\n\tFlash:           16kB");
    printf("\r\n\tRAM:             512B");
    printf("\r\n\tI/O:             16 I/O Pins");
    printf("\r\n\tTimer:           Two 16-bit Timer_A");
    printf("\r\n\tADC:             10-bit 200-ksps");
    printf("\r\n\tComunication:    I2C/SPI/UART");
    printf("\r\n\tDebug interface: JTAG + Spy-Bi-wire");
}

/**************************************************
 * Interrupt Handler
 **************************************************/
__attribute__ ((interrupt(UART_RX_VECTOR)))
void UART_RX_ISR(void)
{
    parameterString[parameterLength] = uart_getc();
    uart_putc(parameterString[parameterLength]); /* Echo */

    /* Also get the characters '\r\n' */
    if (parameterString[parameterLength++] == '\r')
    {
        validCommandFlag = true;
    }
}
